File: frontend/src/features/owner/types/formula.ts
/**
 * Clean domain types for Master Perfumer (Owner) Workstation.
 */

export type FragranceLayer = 'TOP' | 'HEART' | 'BASE' | 'SOLVENT'

export interface FormulaIngredient {
  id?: number
  formula_id?: string
  material_id: string
  material_name: string
  container_code: string
  layer: FragranceLayer
  percentage: number
  target_weight_g: number
  specific_gravity: number
}

export interface Formula {
  id: string
  code: string
  name: string
  version: string
  concentration_pct: number
  concentration_type: string
  maceration_target_days: number
  estimated_cost_per_bottle: number
  is_locked: boolean
  created_by: string
  created_at?: string
  updated_at?: string
  ingredients?: FormulaIngredient[]
}

export interface BatchProjection {
  target_volume_liters: number
  concentration_pct: number
  estimated_maceration_days: number
  evaporation_shrinkage_loss: number
  finished_bottles_50ml: number
  estimated_cost_per_bottle: number
  cost_per_spray_idr: number
  ifra_compliance: string
}

export interface RawMaterialStock {
  id: string
  code: string
  name: string
  category: 'ESSENCE' | 'SOLVENT' | 'BOTTLE' | 'CAP' | 'SPRAYER'
  specific_gravity: number
  stock_grams: number
  cost_per_gram: number
  supplier_name?: string
  lot_number?: string
  coa_number?: string
}

File: backend/internal/models/formula.go
package models

import "time"

type PyramidLayer string

const (
	LayerTop     PyramidLayer = "TOP"
	LayerHeart   PyramidLayer = "HEART"
	LayerBase    PyramidLayer = "BASE"
	LayerSolvent PyramidLayer = "SOLVENT"
)

type FormulaIngredient struct {
	MaterialID      string       `json:"material_id"`
	MaterialName    string       `json:"material_name,omitempty"` // Disembunyikan di Masked Mode
	ContainerCode   string       `json:"container_code"`          // Ditampilkan ke Lab Tech (cth: WADAH-A01)
	Layer           PyramidLayer `json:"layer"`
	Percentage      float64      `json:"percentage"`              // Persentase formula (0-100)
	TargetWeightG   float64      `json:"target_weight_g"`         // Target timbangan gram
	SpecificGravity float64      `json:"specific_gravity"`
}

type Formula struct {
	ID                  string              `json:"id"`
	Code                string              `json:"code"` // cth: FML-SANTAL-01
	Name                string              `json:"name"` // cth: Santal Royale Extrait
	Version             string              `json:"version"` // cth: v2.1
	ConcentrationPct    float64             `json:"concentration_pct"` // cth: 24.0%
	ConcentrationType   string              `json:"concentration_type"` // Extrait / EDP / EDT
	MacerationTargetDays int                `json:"maceration_target_days"`
	Ingredients         []FormulaIngredient `json:"ingredients"`
	EstimatedCostPerBottle float64          `json:"estimated_cost_per_bottle"` // IDR per 50ml
	IsLocked            bool                `json:"is_locked"`
	CreatedBy           string              `json:"created_by"`
	CreatedAt           time.Time           `json:"created_at"`
	UpdatedAt           time.Time           `json:"updated_at"`
}
